moveit_core/planning_scene/src/urdf2collisionobject.cpp

```diff
@@ -0,0 +1,331 @@
+#include <urdf/model.h>
+
+#include <map>
+
+#include <utility>
+
+namespace planning_scene
+{
+/* Auxiliar function, used to compute the direct kinematics
+ * Taken from
+ * https://github.com/ros/robot_state_publisher/blob/ff29b5cdd0c8c1d69c4203de811120ec593e0b2f/src/robot_state_publisher.cpp#L200
+ * */
+void addChildren(const KDL::SegmentMap::const_iterator _segment,
+                 std::map<std::string, robot_state_publisher::SegmentPair>& _segments_fixed);
+
+/* Auxiliar function, used to fix names */
+std::string stripSlash(const std::string& in)
+{
+  if (not in.empty() && in[0] == '/')
+  {
+    return in.substr(1);
+  }
+  return in;
+}
+
+moveit_msgs::CollisionObject urdf_to_collision_object(const urdf::Model& _urdf_model)
+{
+  KDL::Tree kdl_tree;
+  moveit_msgs::CollisionObject result;
+
+  moveit_msgs::CollisionObject result_empty;  // empty collision object should be const
+  // ----------------------------------
+  // -------  Part I ----------
+  //  Get the links from the urdf model and extract its geometry.
+  //  output: the following maps with names and geometries
+  //    - solid_primitives_map
+  //    - mesh_map
+  // ----------------------------------
+  //
+  std::map<std::string, std::pair<shape_msgs::SolidPrimitive, Eigen::Isometry3d>> solid_primitives_map;
+  std::map<std::string, std::pair<shape_msgs::Mesh, Eigen::Isometry3d>> mesh_map;
+
+  kdl_parser::treeFromUrdfModel(_urdf_model, kdl_tree);
+  if (kdl_tree.getSegments().empty())
+  {
+    ROS_ERROR("The urdf is empty");
+    return result_empty;
+  }
+  std::vector<urdf::LinkSharedPtr> links_in_object;
+
+  const std::string& root_link_name = GetTreeElementSegment(kdl_tree.getRootSegment()->second).getName();
+
+  _urdf_model.getLinks(links_in_object);
+
+  std::vector<urdf::LinkSharedPtr>::iterator it =
+      find_if(links_in_object.begin(), links_in_object.end(),
+              [&root_link_name](const urdf::LinkSharedPtr& _link) { return _link->name == root_link_name; });
+
+  if (it == links_in_object.end())
+  {
+    ROS_ERROR("could not find root link");
+    return result_empty;
+  }
+
+  // Place the root object at the beginnig of the array
+  urdf::LinkSharedPtr root_link = *it;
+  links_in_object.erase(it);
+  links_in_object.insert(links_in_object.begin(), root_link);
+
+  shape_msgs::SolidPrimitive solid_primitive;
+
+  for (const urdf::LinkSharedPtr& link : links_in_object)
+  {
+    urdf::GeometrySharedPtr geometry;
+    urdf::Pose pose;
+
+    // Give priority to the collision object description
+    if (link->collision)
+    {
+      geometry = link->collision->geometry;
+      pose = link->collision->origin;
+    }
+    else if (link->visual)
+    {
+      geometry = link->visual->geometry;
+      pose = link->visual->origin;
+    }
+    else
+    {
+      ROS_WARN("Link ill defined: Link %s has not visual or collision tag", link->name.c_str());
+    }
+
+    if (link->collision or link->visual)
+    {
+      switch (geometry->type)
```

Comment:
Converting a link into a pair (primitive, pose) is a good candidate for a separate function as well.